### PR TITLE
Implemented semi-lagrangian advection solver

### PR DESCRIPTION
**Pull request by sdeastham**
The Eulerian advection scheme caused massive diffusion, slow simulations, and had multiple bugs (see #109 and #110). This PR replaces it with a semi-Lagrangian approach, shifting the local vector by floor(CFL) and then performing a single forward Euler step for the remaining fractional timestep. Credit to @coco-yeung for the implementation on which this is based!

This is NOT zero-diff. It dramatically reduces numerical diffusion, which has the side effect of quite severely changing the simulation; in a reference case (issl_rhi140) the lifetime is decreased (see attached). It also increases run speed by 1.2-1.5x in terms of simulation hours per real-time hour.
timeseries_comparison
The reason for this decrease in lifetime seems to be that, with the high degree of numerical diffusion in the Eulerian code, the contrail core is artificially preserved:
extinction_cross_section_0200
A review of the code would be greatly appreciated (nominating @lrobion if possible, and input from @Calebsakhtar would also be greatly appreciated). I believe that this code change does resolve several outstanding issues, but it would be good to verify that the new behaviour is expected.

### CODE REVIEW

**Comment by speth** [comment]
A couple thoughts:
- If we adopt this, I think we need some clear documentation of what the scheme is. This becomes more important as the implementation evolves and neither the Fritz et al (2021) paper nor the Xu (2024) thesis provide a good description of how the code works any more. I'd say the minimal version would be an extended docstring for the `AdvDiffSystem` class, though I realize we don't even build Doxygen docs for this project.
- Could we run this before/after comparison with different grid resolution parameters as a rough convergence study? It would help to see that the two methods are approaching the same solution on a sufficiently fine grid (though perhaps the current method isn't even consistent in the numerical analysis sense, given the known issues).
- If the old scheme was to blame for "slow simulations" but this makes it slower per simulated hour, what does that mean? Of course, correctness beats speed.

**Comment by sdeastham** [comment]
@speth Thanks for this! On each:
1. A docstring makes total sense. I'll add something ASAP. Once the code is a little cleaner I would argue the entire project would benefit from improved documentation, perhaps rising to the level of a model development paper (or at least a public manual).
2. A resolution study is long, long overdue. I've done some thinking on convergence w.r.t. time step - see #126 - but I can extend that to include a grid resolution study. I was a bit reluctant to do so but I see the point that this is necessary.
3. To clarify - this scheme is faster per simulated hour than the old one, but I think the main benefit is really the suppression of numerical diffusion which was causing artificial sublimation of ice crystals.

**Comment by lrobion** [comment]
Hi Seb,
Thanks for this! As a I review this, can you point me to what paper this implementation is based on so that I have a reference. It looks a bit like a flux form semi-Lagrangian approach like Lin & Rood 1996. Is that the one or is there another paper you used?

I agree with Ray that we need new doc, and a before / after (maybe the before should have the fixes to #109 and #110) would be super helpful.

**Comment by sdeastham** [comment]
Thanks @lrobion ! The implementation is much, much simpler than that. Frankly, calling it "semi-Lagrangian" isn't strictly correct but I'm not aware of a better term. The basic idea is:

* Calculate the CFL, c.
* If c exceeds 1, then take the integer part N (e.g. if c = 11.7, take N = 11) and shift the entire vector by N.
* Perform a 1st-order Eulerian flux calculation for a timestep of dt * (c-N)/c.

This only works because the speed is always constant in a given row or column. This is more or less described by Ritchie (1986) (https://doi.org/10.1175/1520-0493(1986)114%3C0135:ETIAWT%3E2.0.CO;2). Hopefully that (somewhat) helps!

Will get started on that doc. I'm on leave next week so that might be a good time.. 😂

**Comment by lrobion** [comment]
I think there might be a small mistake in the forward Euler step because it does not always preserve monotonicity even though we use a flux limiter for that purpose. If you add this test to the test suite this will fail (the first failure being that we create an artificial negative value).

```cpp
// In test_adv_diff_solver.cpp
    TEST_CASE("Semi-Lagrangian 1D Advection preserves monotonicity", "[advection]"){

        // Monotone non-decreasing profile
        const std::vector<double> initial = {0.0, 1.0, 3.0, 4.0, 5.0, 5.0};
        const double dt = 1.0;
        const double ds = 1.0;
        const double bc_left = 0.0;
        const double bc_right = 5.0;

        auto checkMonotonicity = [&](const std::vector<double>& slice){
            // 1. Values should stay in the range spanned by the initial data and the BCs.
            for (std::size_t m = 0; m < slice.size(); m++) {
                INFO("cell " << m << " = " << slice[m]);
                REQUIRE(slice[m] >= 0.0 - 1e-12);
                REQUIRE(slice[m] <= 5.0 + 1e-12);
            }

            // 2. A monotone profile must stay monotone
            for (std::size_t m = 1; m < slice.size(); m++) {
                INFO("cells " << m - 1 << ", " << m << " = " << slice[m-1] << ", " << slice[m]);
                REQUIRE(slice[m] >= slice[m-1] - 1e-12);
            }

            // 3. Total variation must not increase
            double tv_initial = 0.0;
            double tv_final = 0.0;
            for (std::size_t m = 1; m < slice.size(); m++) {
                tv_initial += std::abs(initial[m] - initial[m-1]);
                tv_final += std::abs(slice[m] - slice[m-1]);
            }
            INFO("TV before = " << tv_initial << ", TV after = " << tv_final);
            REQUIRE(tv_final <= tv_initial + 1e-12);
        };

        SECTION("Fractional CFL 0.80, above the 2/3 TVD limit"){
            std::vector<double> slice = initial;
            double velocity = 0.8; // dt = ds = 1 -> no integer shift, fractional CFL = 0.8
            semiLagrangianAdvection1D(slice, velocity, dt, ds, bc_left, bc_right);
            checkMonotonicity(slice);
        }

        SECTION("Fractional CFL 0.60, below the 2/3 TVD limit"){
            std::vector<double> slice = initial;
            double velocity = 0.6; // dt = ds = 1 -> no integer shift, fractional CFL = 0.6
            semiLagrangianAdvection1D(slice, velocity, dt, ds, bc_left, bc_right);
            checkMonotonicity(slice);
        }
    }
```

I think this is because the face flux for cell $m$ e.g. $F_{m + 1/2}$ is incorrectly reconstructed (see line 591 of `AdvDiffSystem.cpp`).

For a cell $m$ assuming positive wind, when we update its value $\phi_m$ with a forward Euler step we do 

$\phi_m^{t+\Delta t} = \phi_m^{t} - c* (F_{m + 1/2} - F_{m - 1/2})$

 which is correct ($c = u\Delta t / \Delta s = CFL)$. But $F_{m + 1/2}$ here represents the time average value of the flux between $t$ and $t + \Delta t$.

We can derive that value by considering the mass that crosses the cell face $m+1/2$ between $t$ and $t + \Delta t$. This mass is exactly $\phi(x)$ for $x \in [x_{m + 1/2 - u\Delta t}, x_{m + 1/2}]$ (mass that gets pushed through by the wind during one timestep).

$m = \int_{x_{m + 1/2 - u\Delta t}}^{x_{m + 1/2}} \phi(x)dx$

Because $c < 1$ by construction (we use forward Euler for the remaining substep) all of this mass lies within a single MUSCL reconstruction (piecewise linear reconstruction of the cell values from the cell average, by definition: $\phi(x) = \phi_m + \sigma_m \xi$ where ,  $\xi = (x − x_m)/ \Delta s$ and $\xi \in [−1/2, 1/2]$ and $\sigma_m$ is the slope we reconstruct (we use minmod in APCEMM).

Then to get $F_{m + 1/2} = m / \Delta t = m / (c * \Delta s)$, we use the previous equation:

 $F_{m + 1/2} =1 / (c \Delta s)  \int_{x_{m + 1/2 - u\Delta t}}^{x_{m + 1/2}} \phi(x)dx$

and substituting $\phi(x)$ with the cell reconstruction + changing variables to $\xi$:

 $F_{m + 1/2} =1 / c  \int_{1/2 - c}^{1/2} (\phi_m + \sigma_m \xi) d\xi$

 $F_{m + 1/2} =1 / c * (c \phi_m + \frac{\sigma_m}{2} [\xi^2]_{1/2 -c}^{1/2}) $

 $F_{m + 1/2} =\phi_m + \frac{\sigma_m}{2} * (1 - c)$

This is different from line 591 (and other places of the SL solver in `AdvDiffSystem.cpp`) which just do $F_{m + 1/2} =\phi_m + \frac{\sigma_m}{2}$ .


Rerunning the test with this fix makes the tests pass. I am not an expert in this so I maybe be wrong, this cropped up because I wanted to add tests for properties we know about the solver here.

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp [hunk 1]
```diff
@@ -541,6 +541,162 @@ namespace FVM_ANDS{
...
+
+                double cfl_frac = abs_vel * rem_dt / ds;
+                for (int m = 0; m < N; ++m) {
+                    slice[m] -= cfl_frac * (face_flux[m] - face_flux[m + 1]);
+                }
+            }
+        }
+    }
+
+    void AdvDiffSystem::semiLagrangianAdvection(double dt) {
```
Comment (lrobion) [inline Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: This should take the `bool parallelAdvection` flag in `FVM_Solver::operatorSplitSolve` and make the `#omp` directive conditional on it otherwise we risk (and we are right now) oversubscribing threads: advection is parallelized at the solver level (see `LAGRIDPlumeModel.cpp` lines 574-588, and now we add a second layer of parallelization within a thread which hurts performance.

The performance gain you saw might be much greater if you ran the benchmark multithreaded.

**Edit**: after doing more reading, OMP may prevent oversubscribing by setting `max_active_levels=1` so this might be inert, but we should still fix it

Comment (lrobion) [reply Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: It also leaves the default parameter `courant_max` (set in the header) of `FVM_Solver::operatorSplitSolve` unused. This could be removed if we fully drop the old solver, but as long as we keep it for comparison purposes we should just have a compile time switch from one to the other or something

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp [hunk 2]
```diff
@@ -541,6 +541,162 @@ namespace FVM_ANDS{
...
+                for (int m = 0; m < N - 1; ++m) {
+                    double diff_up = (m + 1 == N - 1) ? (2.0 * (bc_right - slice[N - 1])) : (slice[m + 2] - slice[m + 1]);
+                    double diff_down = slice[m + 1] - slice[m];
+                    double slope = minmod(diff_down, diff_up);
+                    face_flux[m + 1] = slice[m + 1] - 0.5 * slope;
+                }
+
+                // Outflow face at m = 0
+                double diff_up_0 = 2.0 * (slice[0] - bc_left);
+                double diff_down_0 = (N >= 2) ? (slice[1] - slice[0]) : diff_up_0;
```
Comment (lrobion) [inline Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: I think the `false` branch should be `2.0*(bc_right - slice[0])` mirroring line 595 for the opposite velocity

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp [hunk 3]
```diff
@@ -541,6 +541,162 @@ namespace FVM_ANDS{
...
+                double cfl_frac = abs_vel * rem_dt / ds;
+                for (int m = 0; m < N; ++m) {
+                    slice[m] -= cfl_frac * (face_flux[m] - face_flux[m + 1]);
+                }
+            }
+        }
+    }
+
+    void AdvDiffSystem::semiLagrangianAdvection(double dt) {
+        // 1. Horizontal Advection along X (row by row)
```
Comment (lrobion) [inline Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: It might be worth switching the order of advection x -> y and y -> x in one `operatorSplitSolve` timestep, I think it is supposed to cancel out the bias of doing one first vs the other.

We could add this via a flag and the body of `operatorSplitSolve` would always call the first half-timestep call to the SL solver with x first and the second half-timestep with y first. Or another way to keep track...

I didn't find quantitative justification but it's mentioned in Leonard et al. (1996) "Conservative Explicit Unrestricted-Time-Step Multidimensional Constancy-Preserving Advection Schemes" paragraph below equation 19

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp [hunk 4]
```diff
@@ -541,6 +541,162 @@ namespace FVM_ANDS{
...
+        #pragma omp parallel for default(shared) schedule(static)
+        for (int j = 0; j < ny_; ++j) {
+            double u_j = u_double_ - yCoord_[j] * shear_;
+            if (std::abs(u_j) > 1.0e-14) {
+                std::vector<double> row(nx_);
+                for (int i = 0; i < nx_; ++i) {
+                    int idx = twoDIdx_to_vecIdx(i, j, nx_, ny_, format_);
+                    row[i] = phi_[idx];
+                }
```
Comment (lrobion) [inline Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: Do we expect that these could be empty? I am not 100% sure if an empty boundary conditions vector here is an acceptable code path

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp [hunk 5]
```diff
@@ -541,6 +541,162 @@ namespace FVM_ANDS{
...
+                    }
+                    for (int m = 0; m < k; ++m) {
+                        slice[m] = bc_left;
+                    }
+                }
+            }
+
+            // Step 2: Fractional Forward Euler step
+            if (rem_disp > 1.0e-12 && rem_dt > 0.0) {
+                std::vector<double> face_flux(N + 1, 0.0);
```
Comment (lrobion) [inline Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: We might be able to reduce this allocation (2 (1D solve) x 2 (SL per operatorSplitSolve) x N_bins x N_timestep) with a `static thread_local std::vector<double> face_flux;` at the top of the function + `face_flux.resize(N + 1);` (in place of the allocations) because we always write to this first before reading from it?

Not 100% sure that the gain is large here

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp [hunk 6]
```diff
@@ -541,6 +541,162 @@ namespace FVM_ANDS{
...
+        }
+    }
+
+    void AdvDiffSystem::semiLagrangianAdvection(double dt) {
+        // 1. Horizontal Advection along X (row by row)
+        #pragma omp parallel for default(shared) schedule(static)
+        for (int j = 0; j < ny_; ++j) {
+            double u_j = u_double_ - yCoord_[j] * shear_;
+            if (std::abs(u_j) > 1.0e-14) {
+                std::vector<double> row(nx_);
```
Comment (lrobion) [inline Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: I think we can declare `row` out of the loop in a parallel region such that each threads spawns one buffer they reuse instead of allocating one per iteration: 

```cpp
#pragma omp parallel default(shared)
{
std::vector<double> row(nx_);
#pragma omp for schedule(static)
for (int j = 0; j < ny_; ++j) { ...}
```

This is OMP magic so I am not 100% sure either

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp [hunk 7]
```diff
@@ -541,6 +541,162 @@ namespace FVM_ANDS{
...
+
+                for (int m = 0; m < N - 1; ++m) {
+                    double diff_up = (m + 1 == N - 1) ? (2.0 * (bc_right - slice[N - 1])) : (slice[m + 2] - slice[m + 1]);
+                    double diff_down = slice[m + 1] - slice[m];
+                    double slope = minmod(diff_down, diff_up);
+                    face_flux[m + 1] = slice[m + 1] - 0.5 * slope;
+                }
+
+                // Outflow face at m = 0
+                double diff_up_0 = 2.0 * (slice[0] - bc_left);
```
Comment (lrobion) [inline Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: I think the variable names are swapped, with u < 0, `diff_up_0 = 2.0 * (slice[0] - bc_left);` is downwind and conversely for `diff_down_0`. Minmod is symmetric so I don't think it affects the rest

**Review by lrobion — CHANGES_REQUESTED** [review]
This looks good, I am happy to discuss any thing I've flagged. Some of the allocation stuff is just for performance and I made not sure how much of a difference it would make to hoist them out of the hot loops.

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp (hunk 1 above)
Comment (sdeastham) [reply Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: Very fair. Addressed in https://github.com/MIT-LAE/APCEMM/pull/125/commits/3d7bb3ffdee93e65c3e555381585b92781a736cc.

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp (hunk 2 above)
Comment (sdeastham) [reply Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: This is the left-most boundary condition for a right-to-left flow, so I think slice[0] - bc_left is correct?

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp (hunk 7 above)
Comment (sdeastham) [reply Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: Addressed by 1abc9ce80d749eaa00beb003f3cd52019968e30d.

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp (hunk 4 above)
Comment (sdeastham) [reply Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: I've taken a look through this and don't think that it's possible for these to be empty. I'll edit accordingly to just take the BCs directly.

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp (hunk 6 above)
Comment (sdeastham) [reply Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: Good idea! Attempted this with 1fec6c6471aa11bfd0866d82dfb75c415096caf8 - let me know what you think.

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp (hunk 4 above)
Comment (sdeastham) [reply Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: Took a shot at this as part of 1fec6c6471aa11bfd0866d82dfb75c415096caf8.

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp (hunk 5 above)
Comment (sdeastham) [reply Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: I've opted not to implement this just yet, but let me know if you feel strongly!

**Comment by sdeastham** [comment]
@lrobion - thanks so much for your review! I've tried to make limited-scope edits which I _think_ address the issues you raised, but would value your thoughts.

@speth - there's now a docstring, but the goal is to generate a much more thorough set of documentation which will try to cover recent major changes (including this one). I'll get that resolution study together too, and will make that available for you to review before going any further with the PR.

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp (hunk 5 above)
Comment (lrobion) [reply Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: I don't! We can come back later to do zero-diff optimization changes if we want.

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp (hunk 4 above)
Comment (lrobion) [reply Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: That works for me, I think if the bc vector is empty here it means something went very wrong earlier so failing in an ugly way is fine.

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp (hunk 2 above)
Comment (lrobion) [reply Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: Sorry I may have highlighted the wrong line (with the new variable names):
`double diff_down_0 = 2.0 * (slice[0] - bc_left);` is correct that is the downwind difference (wind going from right to left so downwind is left bc)
but the upwind is incorrect on the `false` branch:
I think the line should be:
`double diff_up_0 = (N >= 2) ? (slice[1] - slice[0]) : 2.0*(bc_right - slice[0]);` because we want the upwind contribution to this cell and upwind here is `bc_right`

Comment (sdeastham) [reply Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: AH! OK - yes, I understand. And while I agree that we should never see this in practice, the code as it stands is incorrect. It also has a good chance of causing misunderstanding of the code's intent! Fixed with 4be1af40417c10cdae0485903c5190fca88260c4.

File: Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp (hunk 3 above)
Comment (sdeastham) [reply Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: I'm reluctant to add this for the moment. While it does make intuitive sense to me, I'm a bit worried because it would mean (in practice) running x -> y -> y -> x in a single time step, which almost feels like it could make things worse? Also, we have an odd case here - our y advection (exclusively ice settling) is basically unrelated to our x (shear-driven) advection. If we had a strong quant justification I'd be totally in support, but given our unusual circumstances I'm hesitant.

Comment (lrobion) [reply Code.v05-00/src/FVM_ANDS/AdvDiffSystem.cpp]: That's fine by me, I don't have a stronger justification for this other than that seems like it would help and the anecdotal mentions of this that I've found. I assume it will not make a very large difference anyway.
We can keep this in mind for the future. but leave it like this for now.

**Comment by lrobion** [comment]
From my end I think we resolved everything for the PR.
Two notes:

1. I think it'd be worth adding the test I commented here as part of the normal test suite
2. optional: if we're in the business of speed there are some optimizations we can do because the hot loops do not vectorize right now (checked with `-fopt-info`) due to branching (ternaries and minmod). But I think transport duration is now dominated by the SOR solve so the gains are likely minor.

**Comment by sdeastham** [comment]
Fantastic - thank you so much @lrobion ! I've added your proposed test (which I'm happy to say the code passes). I'm now working on documentation. I am also performing a resolution sweep, although for that I'm using the code in #126 - it's a PR which will be on top of this one, but which has a couple of additional improvements. They're zero-diff for the standard case though. I'll put the results of the resolution testing there, and I'm happy to hold off on merging both PRs until review is complete for that.

**Review by lrobion — APPROVED** [review]

**Comment by lrobion** [comment]
That works for me. #126 looks very promising, but I'll look into it in detail once you're done with the resolution sweep!

**Review by Calebsakhtar — APPROVED** [review]
Hi @sdeastham this looks great! Looking forward to see how this translates to correctness improvements.

The changes you made in response to what @lrobion suggested seem good, so I just read the review and quickly looked at the rest. I could not spot any obvious bugs and neither could my AI of choice.

My changes are more to do with repeated code, which we probably want to minimise (without sacrificing interpretability). Nevertheless I think that the PR is good to go as-is.

Thanks for doing this!
